eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Plutus/Context.hs

```diff
@@ -0,0 +1,125 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+
+module Cardano.Ledger.Alonzo.Plutus.Context (
+  EraPlutusTxInfo (..),
+  EraPlutusContext (..),
+  mkPlutusLanguageContext,
+
+  -- * Language dependent translation
+  PlutusTxInfo,
+  PlutusTxCert,
+  PlutusScriptPurpose,
+  PlutusScriptContext,
+)
+where
+
+import Cardano.Ledger.Alonzo.Scripts (AlonzoEraScript, PlutusScript (..))
+import Cardano.Ledger.Alonzo.Tx (ScriptPurpose (..))
+import Cardano.Ledger.Binary (DecCBOR (..), EncCBOR (..))
+import Cardano.Ledger.Core
+import Cardano.Ledger.Plutus.Data (Data (..))
+import Cardano.Ledger.Plutus.Language (Language (..), PlutusLanguage)
+import Cardano.Ledger.UTxO (UTxO (..))
+import Cardano.Slotting.EpochInfo (EpochInfo)
+import Cardano.Slotting.Time (SystemStart)
+import Control.DeepSeq (NFData)
+import Data.Kind (Type)
+import Data.Text (Text)
+import NoThunks.Class (NoThunks)
+import qualified PlutusLedgerApi.V1 as P (ToData, toData)
+import qualified PlutusLedgerApi.V1 as PV1
+import qualified PlutusLedgerApi.V2 as PV2
+import qualified PlutusLedgerApi.V3 as PV3
+
+class (PlutusLanguage l, AlonzoEraScript era) => EraPlutusTxInfo (l :: Language) era where
+  toPlutusTxCert :: proxy l -> TxCert era -> Either (ContextError era) (PlutusTxCert l)
```

Comment:
I'm confused about the `ContextError` in these functions. Is it the `data  ContextError era :: Type` defined below in `EraPlutusContext era` class?  I suppose yes, but there needs to be nothing else specified to link the two in that case? Or is that happening when defining the instances? 
If the question doesn't make sense, please ignore it, I will ask you during the meeting tomorrow!

Reply:
I think I might understand the question. I can add `EraPlutusContext` as a superclass to the `EraPlutusTxInfo` if that would make it clearer. However, it is not necessary for a very interesting reason:

Associated data and type families (ones that are defined in a type class) are no different from the standalone data and type families and can in fact be treated the same at the use site. I think the only difference between the two is the fact that you can not create a type family instance without adding a type class instance for the associated type families.

Reply:
Interesting, thank you for the explanation! :+1:

Reply:
I did add `EraPlutusContext` as a superclass, which made a few constraints in one of the testing modules redundant :slightly_smiling_face: